eth/beacon/utils/random.py

```diff
@@ -17,70 +17,80 @@
+            (values_count, RAND_MAX)
         )
 
     output = [x for x in values]
     source = seed
     index = 0
-    while index < values_count:
-        # Re-hash the source
+    while index < values_count - 1:
+        # Re-hash the `source` to obtain a new pattern of bytes.
         source = blake(source)
-        for position in range(0, 30, 3):  # gets indices 3 bytes at a time
-            # Select a 3-byte sampled int
-            sample_from_source = int.from_bytes(source[position:position + 3], 'big')
-            # `remaining` is the size of remaining indices of this round
+
+        # Iterate through the `source` bytes in 3-byte chunks.
+        for position in range(0, 32 - (32 % RAND_BYTES), RAND_BYTES):
+            # Determine the number of indices remaining in `values` and exit
+            # once the last index is reached.
             remaining = values_count - index
-            if remaining == 0:
+            if remaining == 1:
                 break
 
-            # Set a random maximum bound of sample_from_source
-            rand_max = SAMPLE_RANGE - SAMPLE_RANGE % remaining
+            # Read 3-bytes of `source` as a 24-bit big-endian integer.
+            sample_from_source = int.from_bytes(
+                source[position:position + RAND_BYTES], 'big'
+            )
+
+            # Sample values greater than or equal to `sample_max` will cause
+            # modulo bias when mapped into the `remaining` range.
+            sample_max = RAND_MAX - RAND_MAX % remaining
 
-            # Select `replacement_position` with the given `sample_from_source` and `remaining`
-            if sample_from_source < rand_max:
-                # Use random number to get `replacement_position`, where it's not `index`
+            # Perform a swap if the consumed entropy will not cause modulo bias.
+            if sample_from_source < sample_max:
+                # Select a replacement index for the current index.
                 replacement_position = (sample_from_source % remaining) + index
-                # Swap the index-th and replacement_position-th elements
+                # Swap the current index with the replacement index.
                 (output[index], output[replacement_position]) = (
                     output[replacement_position],
                     output[index]
                 )
                 index += 1
             else:
+                # The sample causes modulo bias. A new sample should be read.
                 pass
 
     return output
 
 
 @to_tuple
-def split(seq: Sequence[TItem], pieces: int) -> Iterable[Any]:
+def split(seq: Sequence[TItem], split_count: int) -> Iterable[Any]:
     """
-    Returns the split ``seq`` in ``pieces`` pieces in protocol.
-    Spec: https://github.com/ethereum/eth2.0-specs/blob/0941d592de7546a428066c0473fd1000a7e3e3af/specs/beacon-chain.md#helper-functions  # noqa: E501
+    Returns the split ``seq`` in ``split_count`` pieces in protocol.
```

Comment:
It was pointed out by Piper that there are different ways to do some kind of "splitting a list". So here we noted that this `split` function is required and defined in beacon chain protocol.

Reply:
Wondering if `partition` wouldn't be a better term for this.

Reply:
`partition` might be confusing, because itertoolz already has a [partition](https://toolz.readthedocs.io/en/latest/api.html#toolz.itertoolz.partition) which does something different; clojure [also uses "partition"](https://clojuredocs.org/clojure.core/partition) to refer to the same operation itertoolz uses it for.

Maybe something like: `partition_into_n_pieces`, where the `split_count` parameter is renamed to `n`?

Reply:
This probably isn't the best name but nim calls it [distribute](https://nim-lang.org/docs/sequtils.html#distribute%2Cseq%5BT%5D%2CPositive)

Reply:
@lithp yeah that may be better. `divide` sounds fine to me.